tests/lora/test_lora_layers_peft.py

```diff
@@ -0,0 +1,494 @@
+import os
+import tempfile
+import unittest
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers import CLIPTextModel, CLIPTextModelWithProjection, CLIPTokenizer
+
+from diffusers import (
+    AutoencoderKL,
+    DDIMScheduler,
+    EulerDiscreteScheduler,
+    StableDiffusionPipeline,
+    StableDiffusionXLPipeline,
+    UNet2DConditionModel,
+)
+from diffusers.loaders import AttnProcsLayers
+from diffusers.models.attention_processor import (
+    LoRAAttnProcessor,
+    LoRAAttnProcessor2_0,
+)
+from diffusers.utils.import_utils import is_peft_available
+from diffusers.utils.testing_utils import floats_tensor, require_peft_backend
+
+
+if is_peft_available():
+    from peft import LoraConfig
+    from peft.tuners.tuners_utils import BaseTunerLayer
+    from peft.utils import get_peft_model_state_dict
+
+
+def create_unet_lora_layers(unet: nn.Module):
+    lora_attn_procs = {}
+    for name in unet.attn_processors.keys():
+        cross_attention_dim = None if name.endswith("attn1.processor") else unet.config.cross_attention_dim
+        if name.startswith("mid_block"):
+            hidden_size = unet.config.block_out_channels[-1]
+        elif name.startswith("up_blocks"):
+            block_id = int(name[len("up_blocks.")])
+            hidden_size = list(reversed(unet.config.block_out_channels))[block_id]
+        elif name.startswith("down_blocks"):
+            block_id = int(name[len("down_blocks.")])
+            hidden_size = unet.config.block_out_channels[block_id]
+        lora_attn_processor_class = (
+            LoRAAttnProcessor2_0 if hasattr(F, "scaled_dot_product_attention") else LoRAAttnProcessor
+        )
+        lora_attn_procs[name] = lora_attn_processor_class(
+            hidden_size=hidden_size, cross_attention_dim=cross_attention_dim
+        )
+    unet_lora_layers = AttnProcsLayers(lora_attn_procs)
+    return lora_attn_procs, unet_lora_layers
+
+
+@require_peft_backend
+class PeftLoraLoaderMixinTests:
+    torch_device = "cuda" if torch.cuda.is_available() else "cpu"
+    pipeline_class = None
+    scheduler_cls = None
+    scheduler_kwargs = None
+    has_two_text_encoders = False
+    unet_kwargs = None
+    vae_kwargs = None
+
+    def get_dummy_components(self):
+        torch.manual_seed(0)
+        unet = UNet2DConditionModel(**self.unet_kwargs)
+        scheduler = self.scheduler_cls(**self.scheduler_kwargs)
+        torch.manual_seed(0)
+        vae = AutoencoderKL(**self.vae_kwargs)
+        text_encoder = CLIPTextModel.from_pretrained("ybelkada/tiny-clip-text")
+        tokenizer = CLIPTokenizer.from_pretrained("ybelkada/tiny-clip-text")
+
+        if self.has_two_text_encoders:
+            # text_encoder_2 = CLIPTextModelWithProjection(text_encoder_config)
+            text_encoder_2 = CLIPTextModelWithProjection.from_pretrained("ybelkada/tiny-clip-text-2")
+            tokenizer_2 = CLIPTokenizer.from_pretrained("ybelkada/tiny-clip-text-2")
```

Comment:
Could we leverage the text encoder and the tokenizer from https://huggingface.co/hf-internal-testing/tiny-stable-diffusion-xl-pipe here?

Reply:
This gave me some issues with PEFT since the text encoder is on the subfolder, I preferred to push a separate tiny clip model on the Hub. Happy to move that to another repo

Reply:
ok for me